### PR TITLE
MLDB-1721 pathelements quoted

### PR DESCRIPTION
**Pull request by jeremybarnes**
This patch clarifies what is allowed inside a path element (all characters apart from nulls) and ensures that paths that are written can always be read.  This should fix problems with MLDB accepting control characters in paths on entry, but then failing when processing or printing them.

### CODE REVIEW

File: container_files/public_html/doc/builtin/datasets/Datasets.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
...
+Rows and columns have structured identifiers, called *paths*. A path is an array of UTF-8 strings without null characters called *path elements* (e.g. `['a','b']`) and can be represented as strings called *names* with dots between the path elements (e.g. `'a.b'`). Dots can appear in path elements, causing their string representation to be surrounded by double-quote characters, which can themselves appear in path elements, and are doubled up in the string representation: the path `['a.b', '"hello"']` is stringified as `"a.b"."""hello"""`. These rules make it possible to manipulate hierarchically-structured (i.e. JSON) data within [SQL queries](../sql/Sql.md).
```
Comment (FinchPowers) [inline container_files/public_html/doc/builtin/datasets/Datasets.md]: Is there any mention on whether double quotes are allowed in paths or not?

File: sql/path.cc [hunk 2]
```diff
@@ -447,17 +457,45 @@ toEscapedUtf8String() const
...
+        // Simple character detection doesn't work with UTF-8
+        // Scan it UTF-8 character by UTF-8 character
+        isSimple = true;
+        utf8::iterator<const char *> ufirst(d, d, e);
+        utf8::iterator<const char *> ulast(e, d, e);
+
+        while (isSimple && ufirst != ulast) {
+            auto c = *ufirst++;
+            if (!isSimpleUtf8(c)) {
+                isSimple = false;
```
Comment (FinchPowers) [inline sql/path.cc]: `break;` ?

Comment (jeremybarnes) [inline sql/path.cc]: It breaks because is simple is the loop condition

File: container_files/public_html/doc/builtin/datasets/Datasets.md (hunk 1 above)
Comment (jeremybarnes) [inline container_files/public_html/doc/builtin/datasets/Datasets.md]: Yes in the datasets overview documentation

**Comment by FinchPowers** [comment]
+1

**Pull request merged**